### ISSUE TITLE
Confusing error message when checking generic type variable

### ISSUE THREAD

**Issue by atomicknight** [typetools/checker-framework#594]
Environment: Checker Framework 1.9.11, Oracle JDK 1.8.0_65, Windows 7/8.1

The following class yields a confusing error message:
```java
package org.example;

import org.checkerframework.checker.nullness.qual.Nullable;

public class ResultWrapper<T> {
	private @Nullable T result = null;

	// Should return @Nullable T
	private T getResult() {
		return result;
	}
}
```
Error:
```
d:\workspace\cf-testcases\genericvariable1\src\main\java\org\example\ResultWrapper.java:9: error: [return.type.incompatible] incompatible types in return.
                return result;
                       ^
  found   : T extends @Initialized @Nullable Object
  required: T extends @Initialized @Nullable Object
```

**Comment by mernst** [comment]
Confirmed.  Thanks for the bug report!

I added a test case as file
checker-framework/checker/tests/nullness/generics/GenericReturnField.java

**Comment by smillst** [comment]
Using the `-AprintVerboseGenerics` option makes the found and require types distinct :

````
error: [return.type.incompatible] incompatible types in return.
        return result;
               ^
  found   : T[ extends @Initialized @Nullable Object super @Initialized @Nullable Void]
  required: T[ extends @Initialized @Nullable Object super @Initialized @NonNull Void]
````

This option is needs to be documented in the manual.

**Comment by mernst** [comment]
It is good to document the command-line option.
More importantly, the Checker Framework mechanism for printing "found" and "required" needs to be adjusted so that it prints every component of the types that differ, regardless of the setting of `-AprintVerboseGenerics`.

**Comment by wmdietl** [comment]
The mechanism to print types in error messages already has logic to print "invisible" annotations, if the error message is otherwise identical.
This mechanism should be easy to extend to also automatically use printVerboseGenerics if the error message would otherwise be identical.

**Comment by mernst** [comment]
That mechanism should be used not just if the message would otherwise be identical, but anytime changes would not be visible.  For example, if we have:

 * found: `T[ extends @A super @B]`
 * required: `T[extends @C super @D]`

then it is not enough to print

 * found: `T[ extends @A]`
 * required: `T[extends @C]`

even though those (partial) printed representations are not identical.

**Comment by wmdietl** [comment]
I agree. As soon as one of expected/found is a type variable or wildcard, the verbose generics option should be used. In general, the subtype test might have failed because of the lower bound, which is usually not shown.

**Comment by mernst** [comment]
I don't think we always have to show the full generics information.  In particular, if the hidden information is identical between the two types, then it is just clutter that the user must read through and verify as identical.  But if the hidden information is different, the checker must show it to avoid misleading the user.

**Comment by smillst** [comment]
PR #785 addressed the initial bug submitted by the user; however, it did not address all improvements suggested in https://github.com/typetools/checker-framework/issues/594#issuecomment-183120237.  So this issue will remain open.

**Comment by mernst** [comment]
I feel it might be clearer to close this bug and create a new one just for 
the sub-issue that remains, especially since the issue that remains wasn't 
even part of the original bug report.

**Comment by smillst** [comment]
>PR #785 addressed the initial bug submitted by the user; however, it did not address all improvements suggested in #594 (comment). So this issue will remain open.

A new issue, #788, has been opened to track the remaining improvements.

**Comment by dilraj45** [comment]
I was just wondering why checker framework treats `@Nullable T` and `T` differently. Framework implicitly defaults annotations to `@Nullable` as per the CLIMB-to-top inference rules for generics ([link](https://checkerframework.org/manual/#faq-semantics-section)). Why it treats `@Nullable T` as `T[ extends @Initialized @Nullable Object super @Initialized @Nullable Void]` as `T` as `T[ extends @Initialized @Nullable Object super @Initialized @NonNull Void]`

**Comment by mernst** [comment]
Defaulting never affects a use of a type variable, even if the type variable use has no explicit annotation.  Defaulting helps to choose a single type qualifier for a concrete Java class or interface.  By contrast, a type variable use represents a set of possible types.

**Comment by dilraj45** [comment]
> ```
> package org.example;
> 
> import org.checkerframework.checker.nullness.qual.Nullable;
> 
> public class ResultWrapper<T> {
>     private @Nullable T result = null;
> 
>     // Should return @Nullable T
>     private T getResult() {
>         return result;
>     }
> }
> ```
Apologies, if I'm being silly. The above piece of code still throws the following error message (though with the complete error message). I think this error should not be thrown in the first place. Is it the expected behaviour, should it throw an error and why?
> ```
> error: [return.type.incompatible] incompatible types in return.
>         return result;
>                ^
>   found   : T[ extends @Initialized @Nullable Object super @Initialized @Nullable Void]
>   required: T[ extends @Initialized @Nullable Object super @Initialized @NonNull Void]
Thank you for your help!

**Comment by wmdietl** [comment]
Why should there not be an error?
The return type of `getResult` is `T` and the code returns a `@Nullable T`.
As explained in the manual, these are two different types.
You can see in the error message that one has `@Nullable Void` as super bound, whereas the other has `@NonNull Void` as super bound.
Field `result` is allowed to contain `null`, but `ResultWrapper` can be instantiated with a `@NonNull` type, so returning null would not be sound.
Did you read through the [generics manual section](https://checkerframework.org/manual/#generics)?

**Comment by dilraj45** [comment]
I read through docs, but I was a bit uncertain about the defaulting rules. For as per the [defaulting rules](https://checkerframework.org/manual/#climb-to-top), which says the following

> Implicit upper bounds are defaulted to top to allow them to be instantiated in any way. If a user declared class C\<T\> { ... }, then we assume that the user intended to allow any instantiation of the class, and the declaration is interpreted as class C\<T extends @Nullable Object\> { ... } rather than as class C\<T extends @NonNull Object\> { ... }.

I was under the impression that same applies to the supper bound as well and is implicitly inferred as `@Nullable`. 

Thanks for the clarification!

**Comment by wmdietl** [comment]
Ah, the default lower bound is the bottom qualifier in the hierarchy, because of the same reason: we want to allow all possible instantiations.
Please open a fresh issue that we should document this in the manual.
